### PR TITLE
fix(model-resolution): add free-model fallback for free-only provider setups (fixes #1883)

### LINKED ISSUE #1883: Bug: Should use only free models when no providers configured
## Bug Description

When oh-my-opencode is installed/configured and the user selects no providers, subagents should ONLY use free models. Currently, they incorrectly inherit or use a paid model (glm-4.7) instead of defaulting to free models only.

## Expected Behavior

When no providers are configured:
- Subagents should use ONLY free models (e.g., `opencode/big-pickle`, `opencode/kimi-k2.5-free`, `opencode/minimax-m2.5-free`)
- Should NOT inherit the parent model's configuration
- Should default to free models since no paid providers were set up

## Actual Behavior

- User configures: no providers selected (free models only)
- Main agent may work correctly with a free model
- Subagent (child agent) uses `glm-4.7` (a paid model), causing errors since GLM 4.7 is no longer free
- Subagent should instead use only free models when no providers are configured

## Steps to Reproduce

1. Configure oh-my-opencode with no providers (free models only)
2. Set the main model to a free model like `opencode/kimi-k2.5-free`
3. Spawn a subagent (task with category or subagent_type)
4. Observe that the subagent attempts to use `glm-4.7` instead of a free model

## Environment

- oh-my-opencode version: latest
- Configuration: No providers selected (free models only)

## Additional Context

Reported in: CloudWaddie/actions-agent#67

The fix should ensure that when no providers are configured, the system defaults to using ONLY free models, rather than inheriting any parent model configuration or falling back to a paid model.

### PR DESCRIPTION
**Pull request by mrosnerr**
Fixes #1883

## Summary

Subagents resolve to paid models when only free providers are configured. This adds a free-model fallback that appends free opencode models to the fallback chain when the user only has free models available.

Paid users are unaffected — detection is based on `availableModels`, not `connectedProviders`, so Zen subscribers don't get downgraded.

Free model list is sourced from a generated JSON snapshot refreshed by the existing weekly `refresh-model-capabilities` workflow. Model rotation is a JSON edit, not a code change.

## Changes

| File | Change |
|------|--------|
| `free-model-fallback.ts` | Free-only detection + append strategy |
| `model-selection.ts` | Wire in free-only logic before fallback chain resolution |
| `model-selection-input.ts` | Extract existing helpers (pure refactor) |
| `free-opencode-models.generated.json` | Generated snapshot from models.dev |
| `script/build-free-models.ts` | Standalone build script for the snapshot |
| `script/free-model-extraction.ts` | Extraction logic (testable, separate from runner) |
| `refresh-model-capabilities.yml` | Also runs `build:free-models` on weekly cron |
| `*test.ts` (×3) | Unit, integration, and extraction tests |

## Testing

- `bun run build:free-models` — generates snapshot, verified output matches live `models.dev/api.json`
- `bun run typecheck` — clean
- `bun run build` — clean
- `bun run test:model-capabilities` — 58 pass (includes extraction edge cases + snapshot guardrails)
- All 41 model-selection tests pass, 0 regressions

## Context

Builds on #3529 (closed by @MoerAI, who created the initial module structure).

### CODE REVIEW

File: script/free-model-extraction.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  const models = opencode.models
+  if (!isRecord(models)) return []
+
+  const ids: string[] = []
+  for (const [id, entry] of Object.entries(models)) {
+    if (!isRecord(entry)) continue
+    if (entry.status === "deprecated") continue
+    const cost = entry.cost
+    if (!isRecord(cost)) continue
+    if (cost.input !== 0) continue
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply script/free-model-extraction.ts]: Fixed — now checks both `cost.input` and `cost.output`. Added explicit test cases for each: "excludes models with paid input cost" and "excludes models with paid output cost". Will push shortly.

File: .github/workflows/refresh-model-capabilities.yml [hunk 2]
```diff
@@ -28,18 +28,21 @@ jobs:
       - name: Refresh bundled model capabilities snapshot
```
[the remark this answers is not in the document]
Comment (mrosnerr) [reply .github/workflows/refresh-model-capabilities.yml:28]: This follows the same pattern as the existing `build:model-capabilities` step directly above (line 29) — both fetch from the public `models.dev/api.json` API via `MODELS_DEV_SOURCE_URL`. No repo cloning involved. See [`script/build-model-capabilities.ts`](https://github.com/code-yeongyu/oh-my-openagent/blob/dev/script/build-model-capabilities.ts) for the existing precedent.

**Comment by code-yeongyu** [comment]
[sisyphus-bot]

PR sweep first-pass triage on `dev`.

- Author: `mrosnerr`
- Title: `fix(model-resolution): add free-model fallback for free-only provider setups (fixes #1883)`
- Closing issues: #1883
- Review decision: none yet
- Mergeable state: **CONFLICTING** (rebase needed)
- CI status: green / no failing checks

Needs rebase + review. Please rebase onto current `dev` first; then a review can land cleanly.

Assigning `code-yeongyu` so the maintainer can prioritize a focused review of this PR. If the linked issue above has been triaged, the verdict there is the authoritative direction for this change.

**Comment by code-yeongyu** [comment]
[sisyphus-bot] Hi mrosnerr. 🙏 Thanks for the free-model fallback work for #1883.

Coming back to this from the 5/16 triage queue. Issue #1883 was closed on 5/13 with this PR (#3821) linked as the auto-closer reference, but the PR itself hasn't landed yet, so the underlying behavior is still on the prior path. The PR also currently reports CONFLICTING against current `dev` (11 files in `src/` and `src/cli/` have all moved since 5/06).

If you can spare the time, the cleanest next step is a rebase onto current `dev` (`git fetch upstream && git rebase upstream/dev`), resolving any conflicts in the model-resolution + provider-detection files. Once it's clean I'll run a focused review against the 463-line diff against the latest provider-availability shape; some of the related files (`detect-current-config.ts`, `provider-availability.ts`) were touched by other in-flight PRs this week so the rebase will need to dovetail with those.

If you'd rather close this and re-open a smaller, more focused PR layered on top of the current free-provider detection code, that's also fine. Either way, thanks for caring enough to fix the silent paid-model fallthrough; it's a real footgun for users who deliberately configured only free providers.

**Comment by mrosnerr** [comment]
@code-yeongyu rebased 👍

**Review by MoerAI — CHANGES_REQUESTED** [review]
Thanks for the PR, @mrosnerr, and for caring enough to fix the silent paid-model fallthrough. That is a real footgun for users who deliberately configure only free providers, and the maintainer already flagged the underlying fix as valuable.

The blocker is the same one from the 5/21 pass, now bigger: GitHub marks this as conflicting with `dev`. Since your last rebase the repo completed the `src/` to `packages/omo-opencode/src/` move documented in `AGENTS.md`, so every `src/tools/delegate-task/` and `src/generated/` file this PR touches has changed location. The rebase this time means re-pointing those edits at `packages/omo-opencode/src/tools/delegate-task/` rather than a straight replay.

One smaller note for when you are in there: the change also reaches `.github/workflows/refresh-model-capabilities.yml` and `script/`, which are maintainer-managed areas. If the snapshot and workflow piece can be split from the core free-only detection plus fallback-append, the core fix becomes much easier to review and land on its own.

Once it is rebased clean against current `dev` I will take another look. Thanks again for the work on #1883.

**Comment by mrosnerr** [comment]
@MoerAI Thanks for following up! I've moved in a different direction and won't be able to continue work on this. Please feel free to take this over if you'd like to get it in 🙏

**Comment by MoerAI** [comment]
Thanks for the PR, @mrosnerr, and for the honest update that you have moved on and cannot continue this one.

I am closing it for two reasons. The linked issue #1883 is now CLOSED, and I have not verified its resolution path here, so this fix would need re-confirming against current dev before it could land. The branch has also drifted since June, so a rebase would be needed either way.

Per your note, anyone is welcome to take this over: the cleanest path is a fresh PR against dev that references this one, with a regression test that fails on unmodified dev if the free-only provider bug still reproduces. Happy to reopen if you or another contributor picks it back up and #1883 turns out to still be reproducible. Thanks again for the work you put in.
